src/prefect/futures.py

```diff
@@ -612,19 +613,40 @@ def resolve_futures_to_states(
 
     Unsupported object types will be returned without modification.
     """
-    futures: set[PrefectFuture[R]] = set()
+    return _resolve_futures(
+        expr,
+        resolve_fn=lambda future: future.state if future.wait() or True else None,
+    )
 
-    def _collect_futures(
-        futures: set[PrefectFuture[R]], expr: Any | PrefectFuture[R], context: Any
-    ) -> Any | PrefectFuture[R]:
-        # Expressions inside quotes should not be traversed
-        if isinstance(context.get("annotation"), quote):
-            raise StopVisiting()
 
-        if isinstance(expr, PrefectFuture):
-            futures.add(expr)
+def resolve_futures_to_results(
+    expr: PrefectFuture[R] | Any,
+) -> Any:
+    """
+    Given a Python built-in collection, recursively find `PrefectFutures` and build a
+    new collection with the same structure with futures resolved to their final results.
+    Resolving futures to their final result may wait for execution to complete.
 
-        return expr
+    Unsupported object types will be returned without modification.
+    """
+
+    def _resolve_result(future: PrefectFuture[R]) -> Any:
+        future.wait()
+        if future.state.is_completed():
+            result = future.state.result()
+            return result.result if isinstance(result, ResultRecord) else result
```

Comment:
You should be able to call `.result()` directly on the future here.
```suggestion
            return future.result()
```